hr_employee_benefit/hr_payslip.py

```diff
@@ -0,0 +1,171 @@
+
+        # Field required to compute benefits based on
+        # a percentage of the gross salary for the pay period
+        'gross_salary': fields.float(
+            'Gross Salary',
+            digits_compute=dp.get_precision('Payroll'),
+            readonly=True, states={'draft': [('readonly', False)]},
+        ),
+        # Field required to compute benefits based on an annual amount
+        'pays_per_year': fields.float(
+            'Number of pays per year',
+            digits_compute=dp.get_precision('Payroll'),
+            readonly=True, states={'draft': [('readonly', False)]},
+        ),
+        'benefit_line_ids': fields.one2many(
+            'hr.payslip.benefit.line',
+            'payslip_id',
+            'Employee Benefits',
+            readonly=True, states={'draft': [('readonly', False)]},
+        ),
+    }
+
+    # set_values, compute_benefits and sum_benefits are 3 methods that
+    # can be called directly from inside a salary rule
+
+    def set_values(self, cr, uid, ids, payslip, values, context=None):
+        """
+        Allow to set payslip fields directly from the salary rules.
+
+        This is required to allow setting fields pays_per_year and
+        gross_salary at a precise moment in the salary structure.
+
+        Exemple
+        -------
+        payslip.set_values(payslip, {'gross_salary': 2500, 'pays_per_year':24})
+        """
+        if not isinstance(payslip, orm.browse_record):
+            payslip = payslip.dict
+
+        payslip.write(values)
+
+    def compute_benefits(self, cr, uid, ids, payslip, context=None):
+        """
+        Compute the employee benefits on the payslip.
+
+        This method must be called after the fields gross_salary and
+        pays_per_year are set on the payslip.
+
+        Exemple
+        -------
+        payslip.compute_benefits(payslip)
+        """
+        if not isinstance(payslip, orm.browse_record):
+            payslip = payslip.dict
+
+        payslip.refresh()
+
+        for benefit_line in payslip.benefit_line_ids:
+            if benefit_line.source == 'contract':
+                benefit_line.unlink()
+
+        # Get all employee benefits related to the contract or each job
+        # on the contract
+        contract = payslip.contract_id
+        benefits = contract.benefit_line_ids + list(itertools.chain(*[
+            contract_job.job_id.benefit_line_ids for contract_job
+            in contract.contract_job_ids
+        ]))
+        benefit_ids = [ben.id for ben in benefits]
+
+        # Compute the amounts for each employee benefit
+        self.pool['hr.employee.benefit'].compute_amounts(
+            cr, uid, benefit_ids, payslip, context=context)
+
+    def search_benefits(self, cr, uid, payslip, rule, context=None):
+        """ Get all benefits for a given payslip and a given salary rule
+        :type payslip: hr.payslip browse record
+        :type rule: hr.salary.rule browse record
+        :rtype: list of 
```

Comment:
Trailing white space makes this non-pep8.

Real question, though, is what is this a list of?